pulsar-common/src/main/java/org/apache/pulsar/common/util/collections/SegmentedLongArray.java

```diff
@@ -0,0 +1,121 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.common.util.collections;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.PooledByteBufAllocator;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.concurrent.NotThreadSafe;
+import lombok.Getter;
+
+@NotThreadSafe
+public class SegmentedLongArray implements AutoCloseable {
+
+    private static final int SIZE_OF_LONG = 8;
+
+    private static final int MAX_SEGMENT_SIZE = 2 * 1024 * 1024; // 2M longs -> 16 MB
+    private final List<ByteBuf> buffers = new ArrayList<>();
+
+    @Getter
+    private final long initialCapacity;
+
+    @Getter
+    private long capacity;
+
+    public SegmentedLongArray(long initialCapacity) {
+        long remainingToAdd = initialCapacity;
+
+        while (remainingToAdd > 0) {
+            int sizeToAdd = (int) Math.min(remainingToAdd, MAX_SEGMENT_SIZE);
+            ByteBuf buffer = PooledByteBufAllocator.DEFAULT.directBuffer(sizeToAdd * SIZE_OF_LONG);
+            buffer.writerIndex(sizeToAdd * SIZE_OF_LONG);
+            buffers.add(buffer);
+            remainingToAdd -= sizeToAdd;
+        }
+
+        this.initialCapacity = initialCapacity;
+        this.capacity = this.initialCapacity;
+    }
+
+    public void writeLong(long offset, long value) {
+        int bufferIdx = (int) (offset / MAX_SEGMENT_SIZE);
+        int internalIdx = (int) (offset % MAX_SEGMENT_SIZE);
+        buffers.get(bufferIdx).setLong(internalIdx * SIZE_OF_LONG, value);
+    }
+
+    public long readLong(long offset) {
+        int bufferIdx = (int) (offset / MAX_SEGMENT_SIZE);
+        int internalIdx = (int) (offset % MAX_SEGMENT_SIZE);
+        return buffers.get(bufferIdx).getLong(internalIdx * SIZE_OF_LONG);
+    }
+
+    public void increaseCapacity() {
+        if (capacity < MAX_SEGMENT_SIZE) {
```

Comment:
The size of the last segment might also < MAX_SEGMENT_SIZE, we need to consider to update the capacity of the buffer?

Reply:
And the last segment size < MAX_SEGMENT_SIZE will also affect the `writeLong` and `readLong` method because we always uses the `MAX_SEGMENT_SIZE` to calculate the index.

Reply:
All the segments, after the first one, will be created directly at `MAX_SEGMENT_SIZE` and will not get expanded/shrinked, only added or removed.

Reply:
I think here will add? https://github.com/apache/pulsar/pull/16490/files#diff-ee5391a05253205e8b9bb0f52faa7397326ec0687513204b3bf1b5244d2b8504R45-R50

Reply:
Ah, that's true! I need to fix the constructor

Reply:
Fixed